rfcs/019 - Datastore Compute RFC.md

```diff
@@ -0,0 +1,202 @@
+3. If the remote execution is a "fire-and-forget" mutation on the remote data, the original node will pause execution until it is completed on the other side anyways. (This can be partially worked around through an intermediate data structure that can be more quickly updated, but not eliminated with the proposed opcode.
+
+The first point can be avoided if you can specify whether or not you want a mutable reference to the remote data, the second point if you can specify whether or not you want to use closure semantics or "pure function" semantics (or something in-between), and the third can be avoided if you can specify whether or not you want a return value. With a sufficiently-advanced compiler you could have that all automatically inferred from the way it is being used (is the output assigned to a variable, used in a method chain, operator, etc, or not; is the input argument to the function ever mutated, are any outer-scope variables accessed) but for the first pass, it is proposed that we tackle this with separate explicit opcodes. This is useful even if/when automatic performance improvements to the most-general opcode are handled by the compiler as:
+
+1. It could just emit those well-defined, tighter-bound opcodes and this logic doesn't need to live in the AVM at runtime.
+2. Developers could explicitly opt-in to tighter constraints in "hotter" paths of distributed computing to prevent other developers (or their future self) from accidentally ratcheting down the runtime performance.
+
+Stepping up to the higher-level API design:
+
+```ln
+const baz = ns('someNamespace').ref('foo').run(fn (foo: string) = foo.length());
+```
+
+```ln
+let bar = 'bar';
+const baz = ns('someNamespace').ref('foo').closure(fn (foo: string) {
+  bar = 'foobar: ' + foo + bar;
+  return foo.length();
+});
+```
+
+```ln
+const bar = 'bar';
+const baz = ns('someNamespace').ref('foo').with(bar).run(fn (bar: string, foo: string) = #bar + #foo);
+```
+
+```ln
+const baz = ns('someNamespace').mut('foo').run(fn (foo: string) {
+  foo = foo + 'foo';
+  return foo.length();
+});
+```
+
+```ln
+let bar = 'bar';
+const baz = ns('someNamespace').mut('foo').closure(fn (foo: string) {
+  foo = foo + bar;
+  bar = bar * foo.length();
+  return bar.length();
+});
+```
+
+```ln
+const bar = 'bar';
+const baz = ns('someNamespace').mut('foo').with(bar).run(fn (bar: string, foo: string) {
+  foo = foo * #bar;
+  return foo.length();
+});
+```
+
+```ln
+ns('someNamespace').mut('foo').mutOnly(fn (foo: string) {
+  foo = foo + foo;
+});
+```
+
+```ln
+const bar = 'bar';
+ns('someNamespace').mut('foo').with(bar).mutOnly(fn (bar: string, foo: string) {
+  foo = foo + bar;
+});
+```
+
+All eight potential remote compute possibilities are handled with this API. (Technically the ninth where it's "mutation only" but also mutates the outer scope could exist, but it's identical in performance to `ns().mut().closure()` and so it was elided.)
```

Comment:
But how would we know is a "mutation only" that is mutating an outer scope variable if we use the same `ns().mut().closure()`? Because as I understand it, the `mutOnly` opcode will give us the opportunity to make the mutation necessary but do not stop outer node exec until this inner node finish

Reply:
If we mutate the outer scope then the first node needs to wait for the mutated scope to come back, so it would have to stop outer node execution. That's why I said that it's functionally equivalent.

Reply:
Oh that's true. I got a case in mind but would be handled by the eight case